### PR TITLE
New resources for managing permissions on Areas and Iterations

### PR DESCRIPTION
**Pull request by tmeckel**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

This pull request contains the following new resources:

* `azuredevops_area_permissions` - for managing permissions on areas (components) in Azure DevOps
* `azuredevops_iteration_permissions` - for managing permissions on iterations (sprints) in Azure DevOps

This pull request contains the following new data source:

* `azuredevops_area` - to load information about areas (components) in Azure DevOps
* `azuredevops_iteration` - to load information on iterations (sprints) in Azure DevOps

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Any relevant logs, error output, etc?

(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

### CODE REVIEW

**Comment by tmeckel** [comment]
@xuzhang3 , @EliiseS requesting a review 😁

File: website/docs/d/data_area.html.markdown [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+resource "random_id" "rand_id" {
+  keepers = {
+    seed = var.random-anchor
+  }
+
+  byte_length = 6
+}
+
+locals {
+  project_name = "test-acc-project-${random_id.rand_id.hex}"
```
Comment (xuzhang3) [inline website/docs/d/data_area.html.markdown:24]:
```suggestion
  project_name = "example-project-${random_id.rand_id.hex}"
```

File: website/docs/d/data_area.html.markdown [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+#---------------------------------------------------------------------------
+# Azure DevOps project
+resource "azuredevops_project" "project" {
+  project_name       = local.project_name
+  work_item_template = "Agile"
+  version_control    = "Git"
+  visibility         = "private"
+  description        = "My first project"
```
Comment (xuzhang3) [inline website/docs/d/data_area.html.markdown]:
```suggestion
  description        = "Managed by Terraform"
```

File: website/docs/d/data_area.html.markdown [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    seed = var.random-anchor
+  }
+
+  byte_length = 6
+}
+
+locals {
+  project_name = "test-acc-project-${random_id.rand_id.hex}"
+}
+
```
Comment (xuzhang3) [inline website/docs/d/data_area.html.markdown]: This configuration can be removed. Reduce the configuration complexity for potential users.

File: website/docs/d/data_area.html.markdown [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+#---------------------------------------------------------------------------
+# Azure DevOps project
+resource "azuredevops_project" "project" {
+  project_name       = local.project_name
```
Comment (xuzhang3) [inline website/docs/d/data_area.html.markdown]:
```suggestion
  project_name       = "Sample Project"
```

File: website/docs/d/data_iteration.html.markdown [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  keepers = {
+    seed = var.random-anchor
+  }
+
+  byte_length = 6
+}
+
+locals {
+  project_name = "test-acc-project-${random_id.rand_id.hex}"
+}
```
Comment (xuzhang3) [inline website/docs/d/data_iteration.html.markdown]: This configuration can be removed. Reduce the configuration complexity for potential users.

File: website/docs/d/data_iteration.html.markdown [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+resource "azuredevops_project" "project" {
+  project_name       = local.project_name
+  work_item_template = "Agile"
+  version_control    = "Git"
+  visibility         = "private"
+  description        = "My first project"
```
Comment (xuzhang3) [inline website/docs/d/data_iteration.html.markdown]:
```suggestion
  description        = "Managed by Terraform"
```

File: website/docs/d/data_iteration.html.markdown [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+  byte_length = 6
+}
+
+locals {
+  project_name = "test-acc-project-${random_id.rand_id.hex}"
+}
+
+resource "azuredevops_project" "project" {
+  project_name       = local.project_name
```
Comment (xuzhang3) [inline website/docs/d/data_iteration.html.markdown]:
```suggestion
  project_name       = "Sample Project"
```

File: website/docs/r/area_permissions.html.markdown [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    seed = var.random-anchor
+  }
+
+  byte_length = 6
+}
+
+locals {
+  project_name = "test-acc-project-${random_id.rand_id.hex}"
+}
+
```
Comment (xuzhang3) [inline website/docs/r/area_permissions.html.markdown]: This configuration can be removed. Reduce the configuration complexity for potential users.

File: website/docs/r/area_permissions.html.markdown [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+resource "azuredevops_project" "project" {
+  project_name       = local.project_name
+  work_item_template = "Agile"
+  version_control    = "Git"
+  visibility         = "private"
+  description        = "My first project"
```
Comment (xuzhang3) [inline website/docs/r/area_permissions.html.markdown]:
```suggestion
  description        = "Managed by Terraform"
```

File: website/docs/r/area_permissions.html.markdown [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+  byte_length = 6
+}
+
+locals {
+  project_name = "test-acc-project-${random_id.rand_id.hex}"
+}
+
+resource "azuredevops_project" "project" {
+  project_name       = local.project_name
```
Comment (xuzhang3) [inline website/docs/r/area_permissions.html.markdown]:
```suggestion
  project_name       = "Sample Project"
```

File: website/docs/r/area_permissions.html.markdown [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+	  WORK_ITEM_READ  = "Allow"
+	}
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `project_id` - (Required) The ID of the project to assign the permissions.
```
Comment (xuzhang3) [inline website/docs/r/area_permissions.html.markdown:52]: Can we arrange the parameters with  : Required-> Optional

File: website/docs/r/iteration_permissions.html.markdown [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+## Permission levels
+
+Permission for Iterations within Azure DevOps can be applied on two different levels.
+Those levels are reflected by specifying (or omitting) values for the arguments `project_id` and `path`.
+
+## Example Usage
+
+```hcl
+data "azuredevops_project" "project" {
```
Comment (xuzhang3) [inline website/docs/r/iteration_permissions.html.markdown]: Create a new project is better for example configuration.  Some of the potential users may get confused whit the project not found error.

File: website/docs/r/iteration_permissions.html.markdown [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+	  DELETE          = "Allow"
+	}
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `project_id` - (Required) The ID of the project to assign the permissions.
```
Comment (xuzhang3) [inline website/docs/r/iteration_permissions.html.markdown:62]: Can we arrange the parameters with : Required-> Optional

File: azuredevops/internal/service/workitemtracking/utils/classification_test.go [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+	"github.com/terraform-providers/terraform-provider-azuredevops/azdosdkmocks"
+	"github.com/terraform-providers/terraform-provider-azuredevops/azuredevops/internal/client"
+	"github.com/terraform-providers/terraform-provider-azuredevops/azuredevops/internal/utils/converter"
+	"github.com/terraform-providers/terraform-provider-azuredevops/azuredevops/internal/utils/testhelper"
+)
+
+var classificationProjectName = "test-acc-project-0fu72ecbiu"
+var classificationProjectID = "9c3a5552-268c-423c-a9cd-7de0b36b7035"
+
+type classificationNodeDefintion struct {
```
Comment (xuzhang3) [inline azuredevops/internal/service/workitemtracking/utils/classification_test.go]:
```suggestion
type classificationNodeDefinition struct {
```

File: azuredevops/internal/service/workitemtracking/utils/classification_test.go [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+			children[i] = *convertClassificationTestNode(v)
+		}
+		node.Children = &children
+	} else {
+		node.HasChildren = converter.Bool(false)
+	}
+	return &node
+}
+
+func TestClassifcation_Read_DontSwallowError(t *testing.T) {
```
Comment (xuzhang3) [inline azuredevops/internal/service/workitemtracking/utils/classification_test.go]:
```suggestion
func TestClassification_Read_DontSwallowError(t *testing.T) {
```

File: azuredevops/internal/service/workitemtracking/utils/classification_test.go [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+
+	resourceData := schema.TestResourceDataRaw(t, CreateClassificationNodeSchema(map[string]*schema.Schema{}), nil)
+	resourceData.Set("project_id", classificationProjectID)
+
+	err := ReadClassificationNode(clients, resourceData, workitemtracking.TreeStructureGroupValues.Areas)
+	require.NotNil(t, err)
+	require.Contains(t, err.Error(), errMsg)
+}
+
+func TestClassifcation_Read(t *testing.T) {
```
Comment (xuzhang3) [inline azuredevops/internal/service/workitemtracking/utils/classification_test.go]:
```suggestion
func TestClassification_Read(t *testing.T) {
```

File: azuredevops/internal/service/workitemtracking/utils/classification_test.go [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+	v = resourceData.Get("has_children")
+	require.IsType(t, true, v)
+	require.Equal(t, len(node.children) > 0, v)
+
+	v = resourceData.Get("children")
+	require.NotNil(t, v)
+	require.Len(t, v, len(node.children))
+}
+
+func TestClassifcation_Read_Children(t *testing.T) {
```
Comment (xuzhang3) [inline azuredevops/internal/service/workitemtracking/utils/classification_test.go]:
```suggestion
func TestClassification_Read_Children(t *testing.T) {
```

File: azuredevops/internal/service/workitemtracking/data_iteration_test.go [hunk 18]
```diff
@@ -0,0 +1,12 @@
+package workitemtracking
+
```
Comment (xuzhang3) [inline azuredevops/internal/service/workitemtracking/data_iteration_test.go]: If this  file is not used.  Can we remove it ?

File: azuredevops/internal/service/workitemtracking/data_area_test.go [hunk 19]
```diff
@@ -0,0 +1,12 @@
+package workitemtracking
+
+// The tests in this file use the mock clients in mock_client.go to mock out
```
Comment (xuzhang3) [inline azuredevops/internal/service/workitemtracking/data_area_test.go]: If this  file is not used.  Can we remove it ?

File: azuredevops/internal/service/permissions/resource_area_permissions.go [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+
+	aclToken, err := createAreaToken(clients.Ctx, clients.WorkItemTrackingClient, d)
+	if err != nil {
+		return err
+	}
+
+	err = securityhelper.SetPrincipalPermissions(d, sn, aclToken, nil, false)
+	if err != nil {
+		return err
+	}
```
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/resource_area_permissions.go]:
```suggestion
	if err = securityhelper.SetPrincipalPermissions(d, sn, aclToken, nil, false); err != nil {
		return err
	}
```

File: azuredevops/internal/service/permissions/resource_area_permissions.go [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+		clients.IdentityClient)
+
+	if err != nil {
+		return err
+	}
+
+	aclToken, err := createAreaToken(clients.Ctx, clients.WorkItemTrackingClient, d)
+	if err != nil {
+		return err
+	}
```
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/resource_area_permissions.go]: This code block  has been duplicated times in `CreateOrUpdate`, `Read`, `Delete`,  can you help combine  this code block  into one function?

File: azuredevops/internal/service/permissions/resource_iteration_permissions.go [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+		clients.SecurityClient,
+		clients.IdentityClient)
+	if err != nil {
+		return err
+	}
+
+	aclToken, err := createIterationToken(clients.Ctx, clients.WorkItemTrackingClient, d)
+	if err != nil {
+		return err
+	}
```
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/resource_iteration_permissions.go]: This code block  has been duplicated times in `CreateOrUpdate`, `Read`, `Delete`,  can you help combine  this code block  into one function?

File: azuredevops/internal/service/permissions/resource_iteration_permissions.go [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+
+	err = securityhelper.SetPrincipalPermissions(d, sn, aclToken, nil, false)
+	if err != nil {
+		return err
+	}
```
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/resource_iteration_permissions.go]:
```suggestion
	if err := securityhelper.SetPrincipalPermissions(d, sn, aclToken, nil, false); err != nil {
		return err
	}
```

File: azuredevops/internal/service/permissions/resource_iteration_permissions.go [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+
+	aclToken, err := createIterationToken(clients.Ctx, clients.WorkItemTrackingClient, d)
+	if err != nil {
+		return err
+	}
+
+	err = securityhelper.SetPrincipalPermissions(d, sn, aclToken, &securityhelper.PermissionTypeValues.NotSet, true)
+	if err != nil {
+		return err
+	}
```
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/resource_iteration_permissions.go]:
```suggestion
	if err := securityhelper.SetPrincipalPermissions(d, sn, aclToken, &securityhelper.PermissionTypeValues.NotSet, true); err != nil {
		return err
	}
```

File: azuredevops/internal/service/permissions/resource_area_permissions.go [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+
+	sn, err := securityhelper.NewSecurityNamespace(clients.Ctx,
+		securityhelper.SecurityNamespaceIDValues.CSS,
+		clients.SecurityClient,
+		clients.IdentityClient)
+	if err != nil {
+		return err
+	}
+
+	principalPermissions, err := securityhelper.GetPrincipalPermissions(d, sn, aclToken)
```
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/resource_area_permissions.go:59]: Can we add  `d.SetId("")` for this resource if resource not found ,  like other resources did: https://github.com/terraform-providers/terraform-provider-azuredevops/blob/master/azuredevops/internal/service/taskagent/resource_agent_queue.go#L105

File: azuredevops/internal/service/permissions/resource_iteration_permissions.go [hunk 26]
```diff
@@ -0,0 +1,119 @@
...
+
+	aclToken, err := createIterationToken(clients.Ctx, clients.WorkItemTrackingClient, d)
+	if err != nil {
+		return err
+	}
+
+	principalPermissions, err := securityhelper.GetPrincipalPermissions(d, sn, aclToken)
+	if err != nil {
```
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/resource_iteration_permissions.go:61]: Can we add  `d.SetId("")` for this resource if resource not found ,  like other resources did: https://github.com/terraform-providers/terraform-provider-azuredevops/blob/master/azuredevops/internal/service/taskagent/resource_agent_queue.go#L105

File: website/docs/r/iteration_permissions.html.markdown [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+	principal   = data.azuredevops_group.project-readers.id
+	path        = "Iteration 1"
+	permissions = {
+	  CREATE_CHILDREN = "Allow"
+	  GENERIC_READ    = "NotSet"
+	  DELETE          = "Allow"
+	}
+}
+
+resource "azuredevops_iteration_permissions" "iteration-path-permissions" {
```
Comment (xuzhang3) [inline website/docs/r/iteration_permissions.html.markdown]: Run this example get error :
```
Error: Error getting classification node: VS402485: The Area/Iteration name is not recognized: Iteration 1\Iteration 1_1

  on main.tf line 36, in resource "azuredevops_iteration_permissions" "iteration-path-permissions":
  36: resource "azuredevops_iteration_permissions" "iteration-path-permissions" {

```

**Review by xuzhang3 — CHANGES_REQUESTED** [review]
Hi @tmeckel , New features LGTM,  I left some comments.

File: website/docs/d/data_area.html.markdown (hunk 2 above)
Comment (tmeckel) [reply website/docs/d/data_area.html.markdown]: Done

File: website/docs/d/data_area.html.markdown (hunk 3 above)
Comment (tmeckel) [reply website/docs/d/data_area.html.markdown]: Removed

File: website/docs/d/data_area.html.markdown (hunk 4 above)
Comment (tmeckel) [reply website/docs/d/data_area.html.markdown]: Done

File: website/docs/r/area_permissions.html.markdown (hunk 11 above)
Comment (tmeckel) [reply website/docs/r/area_permissions.html.markdown:52]: Done

File: website/docs/r/iteration_permissions.html.markdown (hunk 12 above)
Comment (tmeckel) [reply website/docs/r/iteration_permissions.html.markdown]: Done

File: azuredevops/internal/service/workitemtracking/data_area_test.go (hunk 19 above)
Comment (tmeckel) [reply azuredevops/internal/service/workitemtracking/data_area_test.go]: Of course

File: website/docs/r/iteration_permissions.html.markdown (hunk 27 above)
Comment (tmeckel) [reply website/docs/r/iteration_permissions.html.markdown]: I removed this HCL from the sample.

File: azuredevops/internal/service/permissions/resource_area_permissions.go (hunk 21 above)
Comment (tmeckel) [reply azuredevops/internal/service/permissions/resource_area_permissions.go]: Refactored

File: azuredevops/internal/service/permissions/resource_iteration_permissions.go (hunk 22 above)
Comment (tmeckel) [reply azuredevops/internal/service/permissions/resource_iteration_permissions.go]: Refactored

File: azuredevops/internal/service/permissions/resource_area_permissions.go (hunk 25 above)
Comment (tmeckel) [reply azuredevops/internal/service/permissions/resource_area_permissions.go:59]: You asked me this already when you were reviewing PR #79 

https://github.com/terraform-providers/terraform-provider-azuredevops/pull/79#discussion_r458646853

And my answer will be here the same as last time: to handle this situation as you wish, I would have to introduce a special error handling.

File: azuredevops/internal/service/permissions/resource_iteration_permissions.go (hunk 26 above)
Comment (tmeckel) [reply azuredevops/internal/service/permissions/resource_iteration_permissions.go:61]: See comment here https://github.com/terraform-providers/terraform-provider-azuredevops/pull/85#discussion_r468461679

**Pull request merged**